src/streams.c

```diff
@@ -84,39 +84,51 @@ static Int READ_COMMAND(Obj *evalResult)
 
 /****************************************************************************
 **
-*F  FuncREAD_ALL_COMMANDS( <self>, <instream>, <echo>, <outputFunc> )
+*F  FuncREAD_ALL_COMMANDS( <self>, <instream>, <echo>, <capture>, <outputFunc> )
 **
-**  FuncREAD_ALL_COMMANDS reads all code from <instream> and returns either
-**  fail if the stream cannot be opened, or executes all statements in the
-**  stream and returns a list of lists, each entry of which reflects the
-**  result of the execution of one statement.
+**  FuncREAD_ALL_COMMANDS attempts to execute all statements read from the
+**  stream <instream> and returns fail if the stream cannot be opened, or a list
+**  of lists, each entry of which reflects the result of the execution of one
+**  statement.
 **
-**  The results are returned as lists of length three if <outputFunc> is
-**  false, or five, if <outputFunc> is a function.
+**  If the parameter <echo> is True, then the statements are echoed to the
+**  current output.
+**
+**  If the parameter <capture> is True, then any output occurring during
+**  execution of a statement, including the output of <outputFunc>, is captured
+**  into a string.
+**
+**  If <resultCallback> is a function, then this function is called on every
+**  statement result, otherwise this parameter is ignored.
+**  Possible outputs of this function are captured if <capture> is True.
+**
+**  The results are returned as lists of length five, the structure of which
+**  is explained below.
+**
+**  The first entry is either true if the statement was executed successfully,
+**  and false otherwise.
 **
 **  If the first entry is true, then the second entry is bound to the result
-**  of the statement if there was one, and unbound otherwise, and the third
-**  entry reflects whether the statement ended in a dual semicolon. If the
-**  first entry is false, an error occurred during the execution of a
-**  statement. In this case, if <outputFunc> is false, only the first entry
-**  is set. If <outputFunc> is a function, only the first and the fifth entry
-**  are set. If <outputFunc> is a function, the fourth entry of the list is
-**  <outputFunc> applied to the result, if there was one, and if the
-**  statement did not end in a dual semicolon. Furthermore, if <outputFunc>
-**  is not false, the fifth entry contains a string of all output printed
-**  while the corresponding statement was executed. If <outputFunc> is false,
-**  all intermediate output is discarded.
+**  of the statement if there was one, and unbound otherwise.
+**
+**  The third entry is true if the statement ended in a dual semicolon.
+**
+**  The fourth entry contains the return value of <resultCallback> if applicable.
+**
+**  The fifth entry contains the captured output as a string, if <capture> is True.
```

Comment:
So the lists are *not* always of length 5. Or rather, they are, but entry 5 may be unbound, which technically is an invalid plist. I guess for low-level code one may just ignore this. But it'd be easy to fix, too, both in this documentation, and in the code: In the docs, change 
> The results are returned as lists of length five
to e.g.
> The results are returned as lists of length at most five
In the code, simply remove `SET_LEN_PLIST(result, 5);` and change all `SET_ELM_PLIST` to `AssPlist`.

I took the liberty of adding a commit which implements all suggested changes (and some more for the docs). Feel free to either drop it again, or to squash it into your previous commit(s) (IMHO this whole PR could be squashed down into a single commit)